apps/mccdb/mccdbcat.1
.Dd April 16, 2024
.Dt mccdbcat 1
.Os mcplex.net
.Sh NAME
.Nm mccdbcat
.Nd dump the contents of a mccurtain AS to IP database file
.Sh SYNOPSIS
.Nm
.Op Fl f Ar dbfile
.Sh DESCRIPTION
.Nm
is a trivial debug utility to dump the contents of a mccurtain AS to IP
database file to stdout in hiuman-readable form.
.Pp

The following command line options are available:
.Pp
.Bl -tag -width indent
.It Fl f Ar dbfile
Specify the input database file to be dumped.  If not given,
/usr/local/etc/mccas2ip.db is assumed.
.El

.Sh FILES
.Bl -tag -width indent
.It Pa /usr/local/etc/mccas2ip.db
The default mccurtain AS to IP database file.
.El
.Sh SEE ALSO
.Xr mccurtain 1 ,
.Xr mccurtaind 8 ,
.Xr mkcurtaindb 8 ,
.Xr credence 1
.Sh AUTHORS
Copyright \(co 2026
.An Daniel W. McRobb Aq Mt dwm@mcplex.net


ChangeLog
2026-09-17  Daniel McRobb  <dwm@mbp22.mcplex.net>

	* tagged as mccurtain-0.0.38

	* apps/mccurtain/Makefile
	* apps/mccurtaind/Makefile
	* apps/mkcurtaindb/Makefile
	* classes/tests/Makefile
	Remove '-c' from compiler arguments when buulding dependencies.

2026-09-17  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.37

	* etc/mccurtaind.cfg.example
	Correct a comment.

2026-09-17  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* tagged as mccurtain-0.0.36

2026-09-17  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/Makefile
	* apps/mccurtaind/mccurtaind.cfg.5
	Added manpage for the mccurtaind configuration file.

2026-09-16  Daniel W. McRobb  <dwm@kiva.mcplex.net>

	* tagged as mccurtain-0.0.35

	* apps/mkcurtaindb/Makefile
	Correct tarprep target path for mccurtaindb.5

	* tagged as mccurtain-0.0.34

2026-09-16  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* apps/mccurtaind/DwmMcCurtainUdpServer.cc
	* apps/mccurtaind/DwmMcCurtainUdpServer.hh
	Check client access in UDP server.

	* etc/mccurtaind.cfg.example
	Add udpAddresses configuration.

	* apps/mkcurtaindb/mccurtaindb.5
	Add manpage for db file format.

	* apps/mccurtaind/mccurtaind.8
	* apps/mccurtaind/Makefile
	Add manpage for mccurtaind(8)

	* apps/mkcurtaindb/mkcurtaindb.8
	* apps/mkcurtaindb/Makefile
	Add manpage for mkcurtaindb(8)

	* tagged as mccurtain-0.0.33

	* classes/include/DwmMcCurtainResponses.hh
	include DwmIpv6Prefix.hh

2026-09-15  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.32

	* apps/mccurtaind/DwmMcCurtainServer.cc
	Populate IPv6 prefixes in responses to AS or country queries.

	* classes/include/DwmMcCurtainResponses.hh
	Add std::vector<Ipv6Prefix> to ASPrefixesResponse so that responses
	to queries for an AS or country can contain IPv6 prefixes.

	* apps/mccurtain/mccurtain.cc
	Added -4 and -6 command line options to select IPv4 or IPv6 prefix
	output when querying for an AS or country.  The default selects only
	IPv4 prefixes.

	* apps/mccurtain/mccurtain.cc
	* apps/mccurtaind/DwmMcCurtainResponder.cc
	* apps/mccurtaind/DwmMcCurtainResponder.hh
	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* classes/include/DwmMcCurtainRequests.hh
	* classes/include/DwmMcCurtainResponses.hh
	Added support for IPv6 address queries via our secure channel.

2026-09-15  Daniel McRobb  <dwm@mbp22.mcplex.net>

	* apps/mccdb/mccdb.cc
	Print country code when printing ASes within a country code.

2026-09-15  Daniel McRobb  <dwm@thrip.mcplex.net>

	* apps/mccdb/mccdb.1
	Added some examples.

	* apps/mccdb/mccdb.cc
	* apps/mccdb/mccdb.1
	Added '-4' and '-6' command line options to select only IPv4 or
	IPv6 prefixes.

2026-09-15  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* Renamed mccdbcat to mccdb.

	* apps/mccdbcat/mccdbcat.cc
	Allow selecting prefixes which match a given address, a given
	country or given AS.  Also allow display of ASes with a
	given country code.

2026-09-13  dwm  <dwm@ria.mcplex.net>

	* tagged as mccurtain-0.0.31

	* apps/mccdbcat/mccdbcat.cc
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/src/DwmMcCurtainDatabaseConfig.cc
	* etc/mccurtaind.cfg.example
	* scripts/updatecurtaindata.sh
	Be consistent about default db filename (now
	/usr/local/etc/mccas2ip.db).

	* tagged as mccurtain-0.0.30

	* apps/mccdbcat/Makefile
	* apps/mccdbcat/mccdbcat.1
	Added manpage for mccdbcat(1)

	* etc/mccurtaind.cfg.example
	Added to comment for dbFile.

	* apps/mccdbcat/mccdbcat.cc
	Use AS2IPDb::Load()

2026-09-13  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/include/DwmMcCurtainAS2IPDb.hh
	* classes/src/DwmMcCurtainAS2IPDb.cc
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/src/Makefile
	* classes/tests/Makefile
	* classes/tests/TestAS2IPDb.cc
	Added AS2IPDb class with static members to load and save
	AS2Ipv4Net and AS2Ipv6Net via common interface that automatically
	uses bz2, gz or no compression based on filename extension.

	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mccurtaind/DwmMcCurtainServer.hh
	Allow loading AS to IP data from bzip2 or gzip files based on
	file extension.

	* apps/mkcurtaindb/mkcurtaindb.cc
	If output file name extension is ".bz2", use BZ2 compression.
	If output file name extension is ".gz", use gzip compression.
	Else use no compression.

2026-09-12  Daniel W. McRobb  <dwm@spark.mcplex.net>

	* classes/include/DwmMcCurtainAS2Ipv4Net.hh
	* classes/src/DwmMcCurtainAS2Ipv4Net.cc
	* classes/tests/TestAS2Ipv4Net.cc
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/tests/TestAS2Ipv6Net.cc
	Save() and Load() will now use bzip2 compression if path extension
	is ".bz2", gzip compression if path extension is ".gz", and no
	compression in all other cases.

2026-09-12  dwm  <dwm@ria.mcplex.net>

	* apps/mccurtaind/DwmMcCurtainDnsServer.hh
	* apps/mccurtaind/DwmMcCurtainServer.cc
	* apps/mkcurtaindb/mkcurtaindb.cc
	* classes/include/DwmMcCurtainAS2Ipv4Net.hh
	* classes/include/DwmMcCurtainAS2Ipv6Net.hh
	* classes/include/DwmMcCurtainIpv4Net2AS.hh
	* classes/include/DwmMcCurtainIpv6Net2AS.hh
	* classes/src/DwmMcCurtainAS2Ipv4Net.cc
	* classes/src/DwmMcCurtainAS2Ipv6Net.cc
	* classes/src/DwmMcCurtainIpv4Net2AS.cc
	* classes/src/DwmMcCurtainIpv6Net2AS.cc
	* classes/tests/.gitignore
	* classes/tests/Makefile
	* classes/tests/TestAS2Ipv4Net.cc
	* classes/tests/TestAS2Ipv6Net.cc
	* classes/tests/TestCaidaV4Routeviews.cc
	* classes/tests/TestCaidaV6Routeviews.cc
	db file now only contains an AS2Ipv4Net and AS2Ipv6Net, since
	we can construct our Ipv4Net2AS and Ipv6Net2AS from them.  This
	reduces thr size of the uncompressed binaery bb file to a size
	less than the gzip'ed CAIDA routeviews files.

	I also added more I/O functionality to AS2Ipv4Net and AS2Ipv6Net

apps/mccdbcat/mccdbcat.cc
//===========================================================================
//  Copyright (c) Daniel W. McRobb 2026
//  All rights reserved.
//
//  Redistribution and use in source and binary forms, with or without
//  modification, are permitted provided that the following conditions
//  are met:
//
//  1. Redistributions of source code must retain the above copyright
//     notice, this list of conditions and the following disclaimer.
//  2. Redistributions in binary form must reproduce the above copyright
//     notice, this list of conditions and the following disclaimer in the
//     documentation and/or other materials provided with the distribution.
//  3. The names of the authors and copyright holders may not be used to
//     endorse or promote products derived from this software without
//     specific prior written permission.
//
//  IN NO EVENT SHALL DANIEL W. MCROBB BE LIABLE TO ANY PARTY FOR
//  DIRECT, INDIRECT, SPECIAL, INCIDENTAL, OR CONSEQUENTIAL DAMAGES,
//  INCLUDING LOST PROFITS, ARISING OUT OF THE USE OF THIS SOFTWARE,
//  EVEN IF DANIEL W. MCROBB HAS BEEN ADVISED OF THE POSSIBILITY OF SUCH
//  DAMAGE.
//
//  THE SOFTWARE PROVIDED HEREIN IS ON AN "AS IS" BASIS, AND
//  DANIEL W. MCROBB HAS NO OBLIGATION TO PROVIDE MAINTENANCE, SUPPORT,
//  UPDATES, ENHANCEMENTS, OR MODIFICATIONS. DANIEL W. MCROBB MAKES NO
//  REPRESENTATIONS AND EXTENDS NO WARRANTIES OF ANY KIND, EITHER
//  IMPLIED OR EXPRESS, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED
//  WARRANTIES OF MERCHANTABILITY OR FITNESS FOR A PARTICULAR PURPOSE,
//  OR THAT THE USE OF THIS SOFTWARE WILL NOT INFRINGE ANY PATENT,
//  TRADEMARK OR OTHER RIGHTS.
//===========================================================================

//---------------------------------------------------------------------------
//!  \file mccurtaindbcat.cc
//!  \author Daniel W. McRobb
//!  \brief trivial utility to dump the contents of the mccurtain db file in
//!    human-readable form
//---------------------------------------------------------------------------

extern "C" {
  #include <unistd.h>
}

#include <cctype>
#include <cstdlib>
#include <fstream>
#include <regex>

#include "DwmMcCurtainAS2IPDb.hh"
#include "DwmMcCurtainIpv4Net2AS.hh"
#include "DwmMcCurtainIpv6Net2AS.hh"
#include "DwmMcCurtainRipeAsnTxt.hh"

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool IsIpv4Address(const std::string & s)
{
  const std::regex
    rgx("(25[0-5]|2[01234][0-9]|1[0-9][0-9]|[0-9][0-9]|[0-9])\\."
        "(25[0-5]|2[01234][0-9]|1[0-9][0-9]|[0-9][0-9]|[0-9])\\."
        "(25[0-5]|2[01234][0-9]|1[0-9][0-9]|[0-9][0-9]|[0-9])\\."
        "(25[0-5]|2[01234][0-9]|1[0-9][0-9]|[0-9][0-9]|[0-9])",
        std::regex::ECMAScript|std::regex::optimize);
  std::smatch  sm;
  return std::regex_match(s, sm, rgx);
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool IsIpv6Address(const std::string & s)
{
  return (Dwm::Ipv6Address(s) != Dwm::Ipv6Address());
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool IsCountryCode(const std::string & s)
{
  if (s.size() == 2) {
    return (std::isupper(s[0]) && std::isupper(s[1]));
  }
  return false;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static bool IsASNumber(const std::string & s)
{
  try {
    auto  u = std::stoul(s);
    return (u <= 0xFFFFFFFF);
  }
  catch (...) {
    return false;
  }
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintAllASes(const Dwm::McCurtain::AS2Ipv4Net & as2ip4,
                         const Dwm::McCurtain::AS2Ipv6Net & as2ip6,
                         const Dwm::McCurtain::RipeAsnTxt & asntxt)
{
  for (const auto & entry : as2ip4.Nets()) {
    std::cout << entry.first;
    auto  asnit = asntxt.Entries().find(entry.first);
    if (asnit != asntxt.Entries().end()) {
      std::cout << ' ' << asnit->second.CountryCode()
                << ' ' << asnit->second.Name();
    }
    std::cout << '\n';
    for (const auto & pfx : entry.second) {
      std::cout << "  " << pfx.first << '\n';
    }
  }
  for (const auto & entry : as2ip6.Nets()) {
    std::cout << entry.first;
    auto  asnit = asntxt.Entries().find(entry.first);
    if (asnit != asntxt.Entries().end()) {
      std::cout << ' ' << asnit->second.CountryCode()
                << ' ' << asnit->second.Name();
    }
    std::cout << '\n';
    for (const auto & pfx : entry.second) {
      std::cout << "  " << pfx.first << '\n';
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintOneAS(const Dwm::McCurtain::AS2Ipv4Net & as2ip4,
                       const Dwm::McCurtain::AS2Ipv6Net & as2ip6,
                       const Dwm::McCurtain::RipeAsnTxt & asntxt,
                       uint32_t asnum)
{
  auto  it4 = as2ip4.Nets().find(asnum);
  bool  printedASNumber = false;
  if (it4 != as2ip4.Nets().end()) {
#if 0
    std::cout << it4->first << '\n';
    printedASNumber = true;
#endif
    for (const auto & pfx : it4->second) {
      std::cout << pfx.first << '\n';
    }
  }
  auto  it6 = as2ip6.Nets().find(asnum);
  if (it6 != as2ip6.Nets().end()) {
#if 0
    if (! printedASNumber) {
      std::cout << it6->first << '\n';
    }
#endif
    for (const auto & pfx : it6->second) {
      std::cout << pfx.first << '\n';
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
template <typename T, typename U>
static void PrintPrefixMatches(const Dwm::McCurtain::RipeAsnTxt & asntxt,
                               const U & addr, const std::vector<T> & matches)
{
  if (! matches.empty()) {
    std::cout << addr << ":\n";
    for (const auto & match : matches) {
      std::cout << "  " << match.first << '\n';
      for (const auto & as : match.second) {
        std::cout << "    " << as;
        auto  asnit = asntxt.Entries().find(as);
        if (asnit != asntxt.Entries().end()) {
          std::cout << ' ' << asnit->second.CountryCode()
                    << ' ' << asnit->second.Name();
        }
        std::cout << '\n';
      }
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintIpv4Matches(const Dwm::McCurtain::AS2Ipv4Net & as2ip4,
                             const Dwm::McCurtain::RipeAsnTxt & asntxt,
                             const Dwm::Ipv4Address & addr)
{
  using Dwm::McCurtain::Ipv4Net2AS;
  Ipv4Net2AS  ip42as;
  if (ip42as.Load(as2ip4)) {
    std::vector<Ipv4Net2AS::value_type>  matches;
    if (ip42as.find_matches(addr, matches)) {
      PrintPrefixMatches(asntxt, addr, matches);
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintIpv6Matches(const Dwm::McCurtain::AS2Ipv6Net & as2ip6,
                             const Dwm::McCurtain::RipeAsnTxt & asntxt,
                             const Dwm::Ipv6Address & addr)
{
  using Dwm::McCurtain::Ipv6Net2AS;
  Ipv6Net2AS  ip62as;
  if (ip62as.Load(as2ip6)) {
    std::vector<Ipv6Net2AS::value_type>  matches;
    if (ip62as.find_matches(addr, matches)) {
      PrintPrefixMatches(asntxt, addr, matches);
    }
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintCountryCodePrefixes(const Dwm::McCurtain::AS2Ipv4Net & as2ip4,
                                     const Dwm::McCurtain::AS2Ipv6Net & as2ip6,
                                     const Dwm::McCurtain::RipeAsnTxt & asntxt,
                                     const std::string & countryCode)
{
  Dwm::McCurtain::CaidaV4Routeviews::ASMapValue  pfxs4;
  
  for (const auto & as : as2ip4.Nets()) {
    auto  asnit = asntxt.Entries().find(as.first);
    if (asnit != asntxt.Entries().end()) {
      if (asnit->second.CountryCode() == countryCode) {
        for (const auto & pfx : as.second) {
          pfxs4.Insert(pfx.first);
        }
      }
    }
  }
  pfxs4.Aggregate();
  std::set<Dwm::Ipv4Prefix>  pfx4set;
  for (const auto & asms : pfxs4.PrefixSets()) {
    for (const auto & pfx : asms.second) {
      pfx4set.insert(pfx);
    }
  }
  for (const auto & pfx : pfx4set) {
    std::cout << pfx << '\n';
  }

  Dwm::McCurtain::CaidaV6Routeviews::ASMapValue  pfxs6;
  
  for (const auto & as : as2ip6.Nets()) {
    auto  asnit = asntxt.Entries().find(as.first);
    if (asnit != asntxt.Entries().end()) {
      if (asnit->second.CountryCode() == countryCode) {
        for (const auto & pfx : as.second) {
          pfxs6.Insert(pfx.first);
        }
      }
    }
  }
  pfxs6.Aggregate();
  std::set<Dwm::Ipv6Prefix>  pfx6set;
  for (const auto & asms : pfxs6.PrefixSets()) {
    for (const auto & pfx : asms.second) {
      pfx6set.insert(pfx);
    }
  }
  for (const auto & pfx : pfx6set) {
    std::cout << pfx << '\n';
  }
  return;
}

//----------------------------------------------------------------------------
//!  
//----------------------------------------------------------------------------
static void PrintCountryCodeASes(const Dwm::McCurtain::RipeAsnTxt & asntxt,
                                 const std::string & countryCode)
{
  std::vector<std::pair<uint32_t,Dwm::McCurtain::RipeAsnTxt::Entry>>  entries;
  for (const auto & asn : asntxt.Entries()) {
    if (asn.second.CountryCode() == countryCode) {
      entries.push_back(asn);
    }
  }
  std::sort(entries.begin(), entries.end(),
            [] (const auto & a, const auto & b)
            { return (a.first < b.first); });
  for (const auto & entry : entries) {
    std::cout << entry.first << ' ' << entry.second.Name() << '\n';
  }
  return;
}

//----------------------------------------------------------------------------
static void Usage(const char *argv0)
{
  std::cerr << "Usage: " << argv0 << " [-f dbfile] [-a asnTxtFile] [-p]\n"
            << "       " << "[ipv4addr|ipv6addr|AS_number|country_code]\n\n"
            << "default dbfile: /usr/local/etc/mccas2ip.db\n"
            << "default asnTxtFile: /usr/local/etc/asn.txt\n";
  return;
}

//----------------------------------------------------------------------------
int main(int argc, char *argv[])
{
  std::string  dbFile("/usr/local/etc/mccas2ip.db");
  std::string  asnTxtFile("/usr/local/etc/asn.txt");

  bool  showPrefixes = false;
  int   optchar;

  while ((optchar = getopt(argc, argv, "a:f:p")) != -1) {
    switch (optchar) {
      case 'a':
        asnTxtFile = optarg;
        break;
      case 'f':
        dbFile = optarg;
        break;
      case 'p':
        showPrefixes = true;
        break;
      default:
        Usage(argv[0]);
        exit(1);
        break;
    }
  }
  
  Dwm::McCurtain::AS2Ipv4Net  as2ipv4;
  Dwm::McCurtain::AS2Ipv6Net  as2ipv6;
  if (Dwm::McCurtain::AS2IPDb::Load(dbFile, as2ipv4, as2ipv6)) {
    Dwm::McCurtain::RipeAsnTxt  asntxt;
    if (asntxt.Load(asnTxtFile)) {
      if (argc > optind) {
        if (IsIpv6Address(argv[optind])) {
          PrintIpv6Matches(as2ipv6, asntxt, Dwm::Ipv6Address(argv[optind]));
        }
        else if (IsIpv4Address(argv[optind])) {
          PrintIpv4Matches(as2ipv4, asntxt, Dwm::Ipv4Address(argv[optind]));
        }
        else if (IsCountryCode(argv[optind])) {
          if (showPrefixes) {
            PrintCountryCodePrefixes(as2ipv4, as2ipv6, asntxt, argv[optind]);
          }
          else {
            PrintCountryCodeASes(asntxt, argv[optind]);
          }
        }
        else if (IsASNumber(argv[optind])) {
          PrintOneAS(as2ipv4, as2ipv6, asntxt, std::stoul(argv[optind]));
        }
      }
      else {
        PrintAllASes(as2ipv4, as2ipv6, asntxt);
      }
    }
    else {
      std::cerr << "Failed to load asn.txt file '" << asnTxtFile << "'\n";
      return 1;
    }
  }
  else {
    std::cerr << "Failed to load db file '" << dbFile << "'\n";
    return 1;
  }
  
  return 0;
}
